.github/workflows/rust.yml

```diff
@@ -178,3 +163,78 @@ jobs:
+  # cargo test and archives JUnit output instead of all of OMICRON_TMP
+  build-and-nextest:
+    env:
+      OMICRON_TMP: /tmp/omicron_tmp
+    needs: skip_duplicate_jobs
+    if: ${{ needs.skip_duplicate_jobs.outputs.should_skip != 'true' }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ ubuntu-18.04, macos-11 ]
+    steps:
+    # actions/checkout@v2
+    - uses: actions/checkout@28c7f3d2b5162b5ddd3dfd9a45aa55eaf396478b
+    - uses: Swatinem/rust-cache@v1
+      if: ${{ github.ref != 'refs/heads/main' }}
+    - uses: taiki-e/install-action@nextest
+    - name: Report cargo version
+      run: cargo --version
+    - name: Remove unnecessary software
+      run: ./tools/ci_clear_space
+    - name: Configure GitHub cache for CockroachDB binaries
+      id: cache-cockroachdb
+      # actions/cache@v2.1.4
+      uses: actions/cache@26968a09c0ea4f3e233fdddbafd1166051a095f6
+      with:
+        key: ${{ runner.os }}-cockroach-binary-${{ hashFiles('tools/cockroachdb_checksums') }}
+        path: "./out/cockroachdb"
+    - name: Configure GitHub cache for ClickHouse binaries
+      id: cache-clickhouse
+      # actions/cache@v2.1.4
+      uses: actions/cache@26968a09c0ea4f3e233fdddbafd1166051a095f6
+      with:
+        key: ${{ runner.os }}-clickhouse-binary-${{ hashFiles('tools/clickhouse_checksums') }}
+        path: "./out/clickhouse"
+    - name: Download ClickHouse
+      if: steps.cache-clickhouse.outputs.cache-hit != 'true'
+      run: ./tools/ci_download_clickhouse
+    - name: Download CockroachDB binary
+      if: steps.cache-cockroachdb.outputs.cache-hit != 'true'
+      run: bash ./tools/ci_download_cockroachdb
+    - name: Create temporary directory for test outputs
+      run: mkdir -p $OMICRON_TMP
+    - name: Build
+      # We build with:
+      # - RUSTFLAGS="-D warnings" RUSTDOCFLAGS="-D warnings": disallow warnings
+      #   in CI builds.  This can result in breakage when the toolchain is
+      #   updated, but that should only happen with a change to the repo, which
+      #   gives us an opportunity to find and fix any newly-introduced warnings.
+      # - `--locked`: do not update Cargo.lock when building.  Checking in
+      #   Cargo.lock ensures that everyone is using the same dependencies and
+      #   also gives us a record of which dependencies were used for each CI
+      #   run.  Building with `--locked` ensures that the checked-in Cargo.lock
+      #   is up to date.
+      # - TMPDIR=$OMICRON_TMP: we specify a specific temporary directory so that
+      #   failed test outputs will be in a known place that we can grab at the
+      #   end without also grabbing random other temporary files.
+      run: |
+        TMPDIR=$OMICRON_TMP PATH="$PATH:$PWD/out/cockroachdb/bin:$PWD/out/clickhouse" RUSTFLAGS="-D warnings" RUSTDOCFLAGS="-D warnings" \
+          cargo build --locked --all-targets --verbose
```

Comment:
I think we could skip the `cargo build` step here, but that might make it harder to tell at a glance when the results of these two jobs diverge